src/util/nextThought.js

```diff
@@ -42,9 +43,15 @@ const isValidContextView = path =>
  * @returns {array | null} rankedContext
  */
 const contextWithThoughtRank = contextInfo => {
```

Comment:
`cursor` contains the rank information, while `contextChain` only contains context. Unfortunately, as long as those data structures are separate I think recreating a `Path` from `contextChain` might be inherently inefficient.

Reply:
> `contextWithThoughtRank` is getting very computationally expensive with `equalArrays` nested in `map`. Are you sure this is necessary? Does the rank need to be recreated at every level, or is only the rank of the head thought used? I think we need to do some deeper analysis of the requirements and performance of this function. Let me know how you'd like to approach that.

We need to have the rank created at every level for the duplicate thoughts. The `setCursor` action might work with incorrect ranks sometimes, as it did in [this case](https://github.com/cybersemics/em/pull/535#discussion_r418649603), but it's important to have the correct rank information to avoid unexpected issues. 
We can improve the performance by 
1) By memorizing the `contextWithThoughtRank` function
2) Albeit not the best way, but using array.join() to compare arrays if it works for our use case

Reply:
> `cursor` contains the rank information, while `contextChain` only contains context. Unfortunately, as long as those data structures are separate I think recreating a `Path` from `contextChain` might be inherently inefficient.

I agree with that. Do you think there's an alternative approach to solve this with the given data structures?

Reply:
> We need to have the rank created at every level for the duplicate thoughts. The setCursor action might work with incorrect ranks sometimes, as it did in this case, but it's important to have the correct rank information to avoid unexpected issues.

Yes, that's true. I was asking because I wasn't sure how the result of `contextWithThoughtRank` was used. If it was used to generate the new `cursor`, then yes it would need the rank at each level. If it was used for some other check, it's possible that it does not need it. I can do a more thorough analysis but I was hoping you could articulate exactly how the result of `contextWithThought` is used so that we can make an informed decision.

> By memorizing [sic] the contextWithThoughtRank function

Memoization is a good choice when you have necessary computation that is called multiple times with the same inputs. In this case I'm arguing that the computation is not necessary. Rather than implement an inefficient algorithm and then memoize it, I suggest optimizing the algorithm first.

> Albeit not the best way, but using array.join() to compare arrays if it works for our use case.

Not a bad idea, although unfortunately it looks like `Array.prototype.join` is considerably slower:
https://jsperf.com/array-string-equality-with-join/

> I agree with that. Do you think there's an alternative approach to solve this with the given data structures?

That is my intuition, but I would like you to investigate this. I have some other priorities I would like to focus on. If you are unable to find a solution, at that point I can allocate some time to investigate more deeply.

I would start by identifying exactly how the result of `contextWithThoughtRank` is used, then look for ways that we already have some of the rank information  or more efficient ways to get it.

Reply:
The result of `contextWithThoughtRank` in `context views` when the cursor down action should traverse the thoughts as if it's a `thought view`. An example - 

```
▼ a 
     ᐁ m
         ∙ x 		
         ▼ b
             ∙ n
             ∙ o
```

When the cursor is at `a.m~.b.n` or further deep down in the context view, the `cursorDown` action needs to work the same way as it does in the `thought view`.If there are any subsequent children (`firstChild`) of the thought in focus, the rank information isn't needed (added in [this commit](https://github.com/cybersemics/em/pull/535/commits/bbd219fb54273a0dd15a65b4f247f854733d29f9)). However, if we need to determine the `nextSibling` or `nextUncle` of the thought, we need the absolute path with rank information, and hence, `contextWithThoughtRank` is called.